ynput/OpenPype
# Global | Nuke: Creator placeholders in workfile template builder

## Brief description
Adding Creator Placeholder plugin implementation

## Description
New placeholders for creators are now available. Templates can be now build included publishable instances.

## Additional info
Several global abstractions at parent workfile building classes had to be altered. Creator plugin abstraction can now be implemented also into other hosts. 

## Need to be tested in:
- [ ] Nuke - new creator feature.
- [ ] Maya - is it working as before and nothing had been broken by this PR?

## Testing notes [ Nuke ]:
1. Test in `\OP02_VFX_demo` ideally on **sh020**  because the template is expecting data from this project. Project can be deployed with `/tools/unpack_project.ps1` (follow instructions inside of the .ps1 file), [here](https://drive.google.com/file/d/1LQkF8r-6q-aD1iSELuHa68JEnAx8MVjb) is where you can download it from.
2. Download [following zip](https://drive.google.com/file/d/1cJCKO8AJzO9Nm7i5-zMMu8D0p2EtDo8Z/view?usp=sharing) and unpack it in project structure as follows.
![image](https://user-images.githubusercontent.com/40640033/209120941-0a8c73d0-6df7-4777-959f-822398ed8115.png)
3. create preset in your test project settings `project_settings/nuke/templated_workfile_build`
4. In case you  already had some preset then `DO NOT SAVE` and test backward compatibility in Nuke by building from template. All should work as it was. 
5.a If you are creating new preset then paste following onto `project_settings/nuke/templated_workfile_build` and fix the path to your in case it is different.
```
{
    "profiles": [
        {
            "task_types": [
                "Compositing"
            ],
            "task_names": [],
            "path": "{root[work]}/{project[name]}/templates/comp.nk",
            "keep_placeholder": false
        }
    ]
}
```
5.b [dowload this settings patch](https://drive.google.com/file/d/1zu6AtXZQ8-35C9YDGoymX9JLXGKmIFxa/view?usp=sharing) then copy context of the file to clipboard. Go to `project_settings/nuke/imageio` and RMB click at the ImageIO and __paste__. It will take about 5mins to update and then save the settings.
6. then open Nuke from sh020 and remove all nodes in scene (just not to collide with already created write node)
![image](https://user-images.githubusercontent.com/40640033/209121413-b1a806c7-7e97-4665-8f33-15822d0a55a0.png)
7. template should show in script and no Placeholders should stay in the scene under loaded Read nodes and Render write node should be created at bottom. Here is how it should look 
![image](https://user-images.githubusercontent.com/40640033/209347035-7b314f16-9c1a-40fb-ad88-cb3649abb38a.png)
9. Try it again but switch the settings `placeholder_keep` to active.
10. Placeholders should stay in the scene after the template is builded. 

## Dependencies:
This PR branch is based on https://github.com/pypeclub/OpenPype/pull/4258 and ideally needs to be merged as second.
